### DISCUSSION TITLE
FAQs on Compilation issues

### DISCUSSION THREAD

**Discussion by CharlesZheZhang** [NCAR/hrldas#72] [Q&A]
We have received many questions related to problems during compiling the HRLDAS/Noah-MP model, many of them are common. 
So we decide to use this discussion page to provide some solutions or guidances to some of the common compilation issues.

**Comment by CharlesZheZhang** [comment]
## 1. LIBJASPER in user_build_options for gfortran compiler

Some users have run into error when compiling the create_forcing.exe with **gfortran** compiler
and the error message looks like this: 

`/usr/bin/gfortran -o create_forcing.exe -I./lib -g -fconvert=big-endian -fbounds-check -fno-range-check -fallow-argument-mismatch -I./lib create_forcing.o \
      -L./lib -lsmda -L/usr/lib -lnetcdf -lnetcdff -lbz2 -ljpeg -L/usr/lib -ljasper
/usr/bin/ld: ./lib/libsmda.a(decode_jpeg2000.o): in function 'info_jpeg2000_':
decode_jpeg2000.c:(.text+0xd5): undefined reference to 'jpc_decode'
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:29: create_forcing.exe] Error 1
make[2]: Leaving directory `

The error message indicates it is related to the jasper library. **First of all make sure the jasper library is installed and the path is set properly.**
 
Then, please refer to this previous issue:
https://github.com/NCAR/hrldas/issues/37#issue-1620342273
 
regarding flags in `LIBJASPER` when compiling create_forcing.exe: 

"the `LIBJASPER` needs to change to: `LIBJASPER = -ljasper -ljpeg`"

If this wouldn't help, please further consult the IT support for your machine.

**Comment by CharlesZheZhang** [comment]
## 2. `fatal error: netcdf.inc: No such file or directory`

This issue would occur if`netcdf.inc` is not in `NETCDFMOD` path when compiling the `module_hrldas_netcdf_io.F` module. This module is related to input/output in hrldas and netcdf library is necessary. 

Make sure that:
1. netcdf library is properly installed
2. use the same compiler to compile both netcdf and hrldas
3. `NETCDFLIB` and `NETCDFMOD` paths are properly set

Here is an example of how to build dependent libraries for WRF, but is similar for hrldas (netcdf and jasper): 
https://www2.mmm.ucar.edu/wrf/OnLineTutorial/compilation_tutorial.php#STEP2

**Comment by CharlesZheZhang** [comment]
## 3. `gfortran: error: unrecognized command line option ‘-fallow-argument-mismatch’; did you mean ‘-Wno-argument-mismatch’?`

This error would occur if you are using an older version of **gfortran** compiler (before gnu/10.1.0). The `F90FLAGS = ... -fallow-argument-mismatch` should change to `-Wno-argument-mismatch`. 

After version gnu/10.1.0, using the `F90FLAGS` in the current `user_build_options` file should be good.

**Comment by CharlesZheZhang** [comment]
## 4. `Program received signal SIGSEGV: Segmentation fault - invalid memory reference.`

Segmentation error is due to memory conflicts. We have received this error from users using gfortran compiler. 

Try adding following options in `F90FLAGS`: 
`F90FLAGS = -g3 -fcheck=all -Wall -fbacktrace ...`

This can trace back to the segfault error when it occurs and hopefully provide more information from the debug symbols from the compile command.

**Reply by NisKarki** [reply]
Hi, has this been fixed? I got this error message while running create_forcing.exe
 Source model land/sea data not found; assuming a bitmap exists

Program received signal SIGSEGV: Segmentation fault - invalid memory reference.

Backtrace for this error:
#0  0x2B9A7F659C37
#1  0x2B9A7F658E30
#2  0x2B9A8030826F
#3  0x442097 in process.3691 at create_forcing.f90:998 (discriminator 21)
#4  0x419A45 in MAIN__ at create_forcing.f90:514 (discriminator 12)
Segmentation fault (core dumped)

**Reply by NisKarki** [reply]
Compiling with intel compiler resolved this as you mentioned, but still let me know if there is a fix for gcc.
